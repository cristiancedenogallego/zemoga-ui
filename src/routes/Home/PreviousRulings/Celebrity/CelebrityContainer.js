import styled from 'styled-components';

const CelebrityContainer = styled.article`
  background-image: linear-gradient(to top, rgba(0,0,0,0.9) 0%, rgba(0,0,0, 0.61) 37%, rgba(0,0,0,0.0) 74%), url(${props => props.imageUrl});
  background-size: cover;
  background-position: 50%;
  display: flex;
  justify-content: flex-end;
  position: relative;
  padding: 35px 40px ${props => `calc(35px + ${props.theme.progressBarHeight})`};
  flex-direction: column;
  min-height: 400px;

  @media (min-width: 700px) {
    min-height: 550px;
  }
`;

export default CelebrityContainer;
