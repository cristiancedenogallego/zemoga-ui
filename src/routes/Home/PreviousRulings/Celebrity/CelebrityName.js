import styled from 'styled-components';

const CelebrityName = styled.div`
  font-size: 45px;
  line-height: 48px;
  color: white;
  margin-bottom: 7px;
`;

export default CelebrityName;
