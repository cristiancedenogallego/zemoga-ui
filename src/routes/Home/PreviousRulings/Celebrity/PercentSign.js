import styled from 'styled-components';

const PercentSign = styled.span`
  font-weight: 300;
  font-size: 20.4px;
  margin-right: 7px;
`;

export default PercentSign;
