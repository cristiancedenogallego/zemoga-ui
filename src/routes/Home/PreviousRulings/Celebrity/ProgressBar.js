import styled, { css } from 'styled-components';
import { rgba } from 'polished';

const ProgressBar = styled.div`
  width: ${props => props.progress}%;
  background: ${props => (props.primary ? rgba(props.theme.colors.primary, 0.7) : rgba(props.theme.colors.secondary, 0.7))};
  justify-content: ${props => (props.primary ? 'flex-start' : 'flex-end')};
  height: 100%;
  box-sizing: border-box;
  ${props => (
    props.progress === 0 ? css`
      display: none;
    ` : 'flex'
  )}
  align-items: center;
  transition: 0.5s ease-in;
  padding: 5px 10px;
`;

export default ProgressBar;
