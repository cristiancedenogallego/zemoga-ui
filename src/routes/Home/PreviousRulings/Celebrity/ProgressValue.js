import styled from 'styled-components';

const ProgressValue = styled.span`
  font-weight: 300;
  font-size: 29.14px;
  margin-left: 7px;
`;

export default ProgressValue;
