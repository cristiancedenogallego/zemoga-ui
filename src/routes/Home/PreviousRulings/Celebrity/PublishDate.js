import styled from 'styled-components';

const PublishDate = styled.time`
  font-weight: bold;
`;

export default PublishDate;
