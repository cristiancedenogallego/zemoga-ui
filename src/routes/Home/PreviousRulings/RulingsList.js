// @flow
import React, { PureComponent } from 'react';
import path from 'ramda/src/path';
import RulingsGrid from './RulingsGrid';
import Celebrity from './Celebrity';

type Props = {};

type State = {
  rulings: Object[],
  isError: boolean,
  votes: Object,
  wasVoted: Object,
}

class RulingsList extends PureComponent<Props, State> {
  state = {
    rulings: [],
    isError: false,
    votes: {},
    wasVoted: {},
  };

  componentDidMount() {
    this.loadData();
    this.loadVotes();
  }

  loadData = async () => {
    try {
      const response = await fetch('/data.json');
      const rulings = await response.json();
      this.setState({
        rulings,
      });
    } catch (ex) {
      this.setState({
        isError: true,
      });
    }
  }

  loadVotes = () => {
    if (typeof Storage !== 'undefined') {
      const jsonVotes = localStorage.getItem('votes');
      const votes = jsonVotes ? JSON.parse(jsonVotes) : {};
      this.setState({
        votes,
      });
    }
  }

  handleVote = (rulingId: number) => (voteValue: -1 | 1) => {
    const { votes } = this.state;
    const nextVotes = { ...votes };
    const currentRulingVotes = nextVotes[rulingId] || {};
    const currentRulingVotesUp = currentRulingVotes.up || 0;
    const currentRulingVotesDown = currentRulingVotes.down || 0;
    const nextRulingVotesUp = voteValue === 1 ? currentRulingVotesUp + 1 : currentRulingVotesUp;
    const nextRulingVotesDown = (
      voteValue === -1 ? currentRulingVotesDown + 1 : currentRulingVotesDown
    );
    const nextRulingVotes = {
      up: nextRulingVotesUp,
      down: nextRulingVotesDown,
    };
    nextVotes[rulingId] = nextRulingVotes;
    localStorage.setItem('votes', JSON.stringify(nextVotes));

    this.setWasVotedRuling(rulingId)(true);
    this.setState({
      votes: nextVotes,
    });
  };

  setWasVotedRuling = (rulingId: number) => (wasVoted: boolean) => {
    this.setState((state) => {
      const nextWasVoted = { ...state.wasVoted };
      nextWasVoted[rulingId] = wasVoted;
      return {
        wasVoted: nextWasVoted,
      };
    });
  };

  render() {
    const { wasVoted, rulings } = this.state;
    return (
      <RulingsGrid>
        {rulings.map(ruling => (
          <Celebrity
            key={ruling.id}
            name={ruling.name}
            description={ruling.description}
            publishedDate={ruling.publishedDate}
            topic={ruling.topic}
            upVotes={path(['votes', ruling.id, 'up'], this.state) || 0}
            downVotes={path(['votes', ruling.id, 'down'], this.state) || 0}
            imageUrl={ruling.imageUrl}
            handleVote={this.handleVote(ruling.id)}
            wasVoted={wasVoted[ruling.id]}
            handleSetWasVotedRuling={this.setWasVotedRuling(ruling.id)}
          />
        ))}
      </RulingsGrid>
    );
  }
}

export default RulingsList;
