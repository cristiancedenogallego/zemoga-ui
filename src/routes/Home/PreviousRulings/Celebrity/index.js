// @flow
import React from 'react';
import moment from 'moment';
import DislikeIcon from '../../../../icons/Dislike';
import LikeIcon from '../../../../icons/Like';
import CelebrityContainer from './CelebrityContainer';
import CelebrityName from './CelebrityName';
import RatingButton from './RatingButton';
import ProgressValue from './ProgressValue';
import ProgressBar from './ProgressBar';
import PercentSign from './PercentSign';
import RatingBar from './RatingBar';
import PublishInfo from './PublishInfo';
import PublishDate from './PublishDate';
import VoteView from './VoteView';
import WasVotedMessage from './WasVotedMessage';

type Props = {
  name: string,
  description: string,
  publishedDate: string,
  topic: string,
  upVotes: number,
  downVotes: number,
  imageUrl: string,
  wasVoted: boolean,
  handleVote: Function,
  handleSetWasVotedRuling: Function,
};

const RatingStatus = RatingButton.withComponent('div').extend`
  position: absolute;
  bottom: 230px;
  left: 0;
`;

function Celebrity({
  downVotes,
  upVotes,
  imageUrl,
  name,
  publishedDate,
  topic,
  wasVoted,
  handleSetWasVotedRuling,
  description,
  handleVote,
}: Props) {
  const totalVotes = downVotes + upVotes;
  const percentPositiveVotes = Math.round(totalVotes > 0 ? (upVotes * 100) / totalVotes : 0);
  const isGoodRating = percentPositiveVotes > 50;
  const percentNegativeVotes = Math.round(totalVotes > 0 ? 100 - percentPositiveVotes : 0);
  return (
    <CelebrityContainer imageUrl={imageUrl}>
      <CelebrityName>{name}</CelebrityName>
      <PublishInfo>
        <PublishDate>
          {moment(publishedDate).fromNow()}
        </PublishDate>
        <span>In </span>
        <span>{topic}</span>
      </PublishInfo>
      {wasVoted && (
        <WasVotedMessage handleSetWasVotedRuling={handleSetWasVotedRuling} />
      )}
      {!wasVoted && (
        <VoteView description={description} handleVote={handleVote} />
      )}
      <RatingStatus isGoodRating={isGoodRating}>
        {isGoodRating && <LikeIcon width="18px" />}
        {!isGoodRating && <DislikeIcon width="18px" />}
      </RatingStatus>
      <RatingBar>
        <ProgressBar progress={percentPositiveVotes} primary>
          <LikeIcon width="28px" />
          <ProgressValue>{percentPositiveVotes}</ProgressValue>
          <PercentSign>%</PercentSign>
        </ProgressBar>
        <ProgressBar progress={percentNegativeVotes} secondary>
          <ProgressValue>{percentNegativeVotes}</ProgressValue>
          <PercentSign>%</PercentSign>
          <DislikeIcon width="28px" />
        </ProgressBar>
      </RatingBar>
    </CelebrityContainer>
  );
}

export default Celebrity;
