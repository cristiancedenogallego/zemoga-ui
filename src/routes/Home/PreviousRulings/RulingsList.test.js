import React from 'react';
import { mount, shallow } from 'enzyme';
import RulingsList from './RulingsList';
import Celebrity from './Celebrity';

jest.mock('./Celebrity', () => 'Celebrity');

const fetchSpy = jest.spyOn(window, 'fetch');
const stubList = [{ a: 1 }];
const rulingsPromise = Promise.resolve(stubList);
const stubResponse = Promise.resolve({ json: () => rulingsPromise });

fetchSpy.mockReturnValue(stubResponse);

describe('RulingsList', () => {
  test('Should load data from proper .json file', () => {
    mount(
      <RulingsList />,
    );
    expect(fetchSpy).toHaveBeenCalledWith('/data.json');
  });

  test('Should update state from fetch response', async () => {
    const wrapper = mount(
      <RulingsList />,
    );
    await rulingsPromise;
    await stubResponse;
    expect(wrapper.state().rulings).toBe(stubList);
  });

  test('Should update state on fetch fails', async () => {
    const stubFetchFail = Promise.reject();
    fetchSpy.mockReturnValueOnce(stubFetchFail);
    const wrapper = mount(
      <RulingsList />,
    );
    try {
      await stubFetchFail;
    } catch (ex) {
      expect(wrapper.state().isError).toBe(true);
    }
  });

  test('Should render proper Celebrity cards', () => {
    const wrapper = shallow(
      <RulingsList />,
    );
    wrapper.setState({
      rulings: [
        {
          id: 1,
          name: 'Kanye West',
          description: 'Vestibulum diam ante, porttitor a odio eget, rhoncus neque. Aenean eu velit libero.',
          publishedDate: '2018-07-01',
          topic: 'Entertainment',
          imageUrl: '/images/kanye_west.jpg',
        },
      ],
    });
    wrapper.update();
    expect(wrapper.find(Celebrity).props()).toMatchSnapshot();
  });
});
