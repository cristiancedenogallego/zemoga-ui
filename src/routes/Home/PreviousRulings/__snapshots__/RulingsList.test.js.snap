// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RulingsList Should render proper Celebrity cards 1`] = `
Object {
  "description": "Vestibulum diam ante, porttitor a odio eget, rhoncus neque. Aenean eu velit libero.",
  "downVotes": 0,
  "handleSetWasVotedRuling": [Function],
  "handleVote": [Function],
  "imageUrl": "/images/kanye_west.jpg",
  "name": "Kanye West",
  "publishedDate": "2018-07-01",
  "topic": "Entertainment",
  "upVotes": 0,
  "wasVoted": undefined,
}
`;
